jni/include/native_engines_stream_support.h

```diff
@@ -119,6 +120,105 @@ class NativeEngineIndexInputMediator {
 
 
 
+/**
+ * This class delegates the provided index output to do IO processing.
+ * In most cases, it is expected that IndexOutputWithBuffer was passed down to this,
+ * which eventually have Lucene's IndexOutput to write bytes.
+ */
+class NativeEngineIndexOutputMediator {
+ public:
+  NativeEngineIndexOutputMediator(JNIUtilInterface *_jni_interface,
+                                  JNIEnv *_env,
+                                  jobject _indexOutput)
+      : jni_interface(_jni_interface),
+        env(_env),
+        indexOutput(_indexOutput),
+        bufferArray((jbyteArray) (_jni_interface->GetObjectField(_env,
+                                                                 _indexOutput,
+                                                                 getBufferFieldId(_jni_interface, _env)))),
+        writeBytesMethod(getWriteBytesMethod(_jni_interface, _env)),
+        bufferLength(jni_interface->GetJavaBytesArrayLength(env, bufferArray)),
+        nextWriteIndex() {
```

Comment:
Should we explicitly initialize this?

Reply:
I thought it would be better than otherwise!
Caching the values to avoid calling methods several times, and initialize 0 to `nextWriteIndex`.
The mediator will be allocated in stack area, having few bytes and inlining them should be better. Also it will give more readability!